distributed/scheduler.py

```diff
@@ -4893,8 +4881,31 @@ async def update_graph(
                     dask.order.order, dsk=dsk, dependencies=stripped_deps
                 )
             ordering_done = time()
+
             logger.debug("Ordering done.")
 
+            # *************************************
+            # BELOW THIS LINE HAS TO BE SYNCHRONOUS
+            #
+            # Everything that compares the submitted graph to the current state
+            # has to happen in the same event loop.
+            # *************************************
+
+            lost_keys = self._find_lost_dependencies(dsk, dependencies, keys)
+
+            if lost_keys:
+                self.report(
+                    {
+                        "op": "cancelled-keys",
+                        "keys": lost_keys,
+                        "reason": "lost dependencies",
+                    },
+                    client=client,
+                )
+                self.client_releases_keys(
+                    keys=lost_keys, client=client, stimulus_id=stimulus_id
+                )
+
```

Comment:
This is a nasty one and I think I'll want this in another PR. the lost keys check is currently performed between materialization and ordering (for no reason, just chance) which are both offloaded to give the event loop a chance to breath.
However, that also means that tasks can be cancelled during the ordering step that would otherwise cause the computation to be flagged as lost. Instead of receiving a CancelledError the user will see obscure state transition issues.